File: api/budget_bid_conversion.py
"""Phase 4 budget-to-bid conversion sessions and immutable lineage."""
from __future__ import annotations
import json
from datetime import datetime, timezone
from uuid import uuid4
from flask import Blueprint, jsonify, request
from sqlalchemy import Column, DateTime, Integer, MetaData, String, Table, Text, and_, select

metadata = MetaData()
conversion_sessions = Table(
    "budget_bid_conversion_sessions", metadata,
    Column("id", String(100), primary_key=True),
    Column("source_project_code", String(100), nullable=False, index=True),
    Column("source_budget_version_id", String(100), nullable=False),
    Column("target_bid_project_code", String(100), nullable=False, index=True),
    Column("mode", String(20), nullable=False),
    Column("status", String(20), nullable=False),
    Column("options_json", Text, nullable=False),
    Column("source_snapshot_json", Text, nullable=False),
    Column("result_snapshot_json", Text, nullable=False),
    Column("created_by", String(100), nullable=False),
    Column("created_at", DateTime(timezone=True), nullable=False),
    Column("row_version", Integer, nullable=False, default=1),
)

class BudgetBidConversionService:
    def __init__(self, engine):
        self.engine = engine
        metadata.create_all(engine)

    def convert(self, body: dict, actor: str) -> dict:
        source_project = str(body.get("source_project_code", "")).strip()
        source_version = str(body.get("source_budget_version_id", "")).strip()
        target_project = str(body.get("target_bid_project_code", "")).strip()
        mode = str(body.get("mode", "CREATE")).strip().upper()
        if not source_project or not source_version or not target_project:
            raise ValueError("source project, source version and target bid project are required")
        if mode not in {"CREATE", "REPLACE", "APPEND"}:
            raise ValueError("mode must be CREATE, REPLACE or APPEND")
        source_items = list(body.get("budget_items") or [])
        if not source_items:
            raise ValueError("budget_items are required")
        converted = []
        seen = set()
        for index, raw in enumerate(source_items):
            item_id = str(raw.get("id", "")).strip() or f"ROW-{index+1}"
            if item_id in seen:
                raise ValueError("duplicate source budget item id")
            seen.add(item_id)
            converted.append({
                "source_budget_item_id": item_id,
                "bid_item_id": f"{target_project}:{item_id}",
                "code": str(raw.get("code", "")).strip().upper(),
                "name": str(raw.get("name", "")).strip(),
                "unit": str(raw.get("unit", "")).strip(),
                "quantity": str(raw.get("quantity", "0")),
                "unit_price": str(raw.get("unit_price", "0")),
                "amount": str(raw.get("amount", "0")),
                "sort_order": int(raw.get("sort_order", index + 1)),
            })
        converted.sort(key=lambda item: (item["sort_order"], item["code"], item["source_budget_item_id"]))
        session_id = str(uuid4())
        now = datetime.now(timezone.utc)
        options = dict(body.get("options") or {})
        with self.engine.begin() as conn:
            if mode == "CREATE" and conn.execute(select(conversion_sessions.c.id).where(
                conversion_sessions.c.target_bid_project_code == target_project
            )).first():
                raise RuntimeError("CONFLICT")
            conn.execute(conversion_sessions.insert().values(
                id=session_id, source_project_code=source_project,
                source_budget_version_id=source_version,
                target_bid_project_code=target_project, mode=mode, status="COMPLETED",
                options_json=json.dumps(options, ensure_ascii=False, sort_keys=True),
                source_snapshot_json=json.dumps(source_items, ensure_ascii=False, sort_keys=True),
                result_snapshot_json=json.dumps(converted, ensure_ascii=False, sort_keys=True),
                created_by=actor, created_at=now, row_version=1,
            ))
        return self.get(session_id)

    def get(self, session_id: str) -> dict:
        with self.engine.connect() as conn:
            row = conn.execute(select(conversion_sessions).where(conversion_sessions.c.id == session_id)).mappings().first()
        if not row:
            raise LookupError("conversion session not found")
        item = dict(row)
        item["options"] = json.loads(item.pop("options_json"))
        item["source_snapshot"] = json.loads(item.pop("source_snapshot_json"))
        item["result_snapshot"] = json.loads(item.pop("result_snapshot_json"))
        item["created_at"] = item["created_at"].isoformat()
        item["lineage"] = {
            "source_project_code": item["source_project_code"],
            "source_budget_version_id": item["source_budget_version_id"],
            "target_bid_project_code": item["target_bid_project_code"],
            "session_id": item["id"],
        }
        item["deep_link"] = f"/app/bid-conversion?session={item['id']}"
        return item


def build_budget_bid_conversion_blueprint(service: BudgetBidConversionService, resolve_user_id):
    bp = Blueprint("budget_bid_conversion", __name__, url_prefix="/api/conversions")
    @bp.post("/budget-to-bid")
    def convert():
        actor = resolve_user_id()
        if actor is None:
            return jsonify({"code": "UNAUTHORIZED"}), 401
        try:
            return jsonify(service.convert(request.get_json(silent=True) or {}, str(actor)))
        except ValueError as exc:
            return jsonify({"code": "INVALID_ARGUMENT", "detail": str(exc)}), 400
        except RuntimeError:
            return jsonify({"code": "CONFLICT", "detail": "target bid project already has a conversion"}), 409
    @bp.get("/sessions/<session_id>")
    def get_session(session_id: str):
        try:
            return jsonify(service.get(session_id))
        except LookupError as exc:
            return jsonify({"code": "NOT_FOUND", "detail": str(exc)}), 404
    return bp

File: pcces-go/internal/storage/sqlite/budget_bid_conversion.go
package sqlite

import (
	"context"
	"database/sql"
	"encoding/json"
	"fmt"
	"sort"
	"strings"
	"time"

	errx "github.com/liuxb99/pcces-web/pcces-go/internal/platform/errors"
)

type BudgetBidConversionItem struct {
	SourceBudgetItemID string `json:"source_budget_item_id"`
	BidItemID          string `json:"bid_item_id"`
	Code               string `json:"code"`
	Name               string `json:"name"`
	Unit               string `json:"unit"`
	Quantity           string `json:"quantity"`
	UnitPrice          string `json:"unit_price"`
	Amount             string `json:"amount"`
	SortOrder          int    `json:"sort_order"`
}

type BudgetBidConversionRequest struct {
	SourceProjectCode     string           `json:"source_project_code"`
	SourceBudgetVersionID string           `json:"source_budget_version_id"`
	TargetBidProjectCode  string           `json:"target_bid_project_code"`
	Mode                  string           `json:"mode"`
	ActorID               string           `json:"actor_id"`
	Options               map[string]any   `json:"options"`
	BudgetItems           []map[string]any `json:"budget_items"`
}

type BudgetBidConversionSession struct {
	ID                    string                    `json:"id"`
	SourceProjectCode     string                    `json:"source_project_code"`
	SourceBudgetVersionID string                    `json:"source_budget_version_id"`
	TargetBidProjectCode  string                    `json:"target_bid_project_code"`
	Mode                  string                    `json:"mode"`
	Status                string                    `json:"status"`
	Options               map[string]any            `json:"options"`
	SourceSnapshot        []map[string]any          `json:"source_snapshot"`
	ResultSnapshot        []BudgetBidConversionItem `json:"result_snapshot"`
	CreatedBy             string                    `json:"created_by"`
	CreatedAt             string                    `json:"created_at"`
	RowVersion            int64                     `json:"row_version"`
	Lineage               map[string]string         `json:"lineage"`
	DeepLink              string                    `json:"deep_link"`
}

type BudgetBidConversionRepository struct{ store *Store }

func NewBudgetBidConversionRepository(store *Store) *BudgetBidConversionRepository {
	return &BudgetBidConversionRepository{store: store}
}

func (r *BudgetBidConversionRepository) Convert(ctx context.Context, req BudgetBidConversionRequest) (BudgetBidConversionSession, error) {
	req.SourceProjectCode = strings.TrimSpace(req.SourceProjectCode)
	req.SourceBudgetVersionID = strings.TrimSpace(req.SourceBudgetVersionID)
	req.TargetBidProjectCode = strings.TrimSpace(req.TargetBidProjectCode)
	req.Mode = strings.ToUpper(strings.TrimSpace(req.Mode))
	if req.Mode == "" {
		req.Mode = "CREATE"
	}
	if req.SourceProjectCode == "" || req.SourceBudgetVersionID == "" || req.TargetBidProjectCode == "" {
		return BudgetBidConversionSession{}, errx.New(errx.CodeInvalidArgument, "source project, source version and target bid project are required", "P4-CONV-005")
	}
	if req.Mode != "CREATE" && req.Mode != "REPLACE" && req.Mode != "APPEND" {
		return BudgetBidConversionSession{}, errx.New(errx.CodeInvalidArgument, "mode must be CREATE, REPLACE or APPEND", "P4-CONV-005")
	}
	if len(req.BudgetItems) == 0 {
		return BudgetBidConversionSession{}, errx.New(errx.CodeInvalidArgument, "budget_items are required", "P4-CONV-005")
	}
	seen := map[string]bool{}
	converted := make([]BudgetBidConversionItem, 0, len(req.BudgetItems))
	for i, raw := range req.BudgetItems {
		id := strings.TrimSpace(fmt.Sprint(raw["id"]))
		if id == "" || id == "<nil>" {
			id = fmt.Sprintf("ROW-%d", i+1)
		}
		if seen[id] {
			return BudgetBidConversionSession{}, errx.New(errx.CodeInvalidArgument, "duplicate source budget item id", "P4-CONV-005")
		}
		seen[id] = true
		converted = append(converted, BudgetBidConversionItem{
			SourceBudgetItemID: id,
			BidItemID:          req.TargetBidProjectCode + ":" + id,
			Code:               strings.ToUpper(strings.TrimSpace(fmt.Sprint(raw["code"]))),
			Name:               strings.TrimSpace(fmt.Sprint(raw["name"])),
			Unit:               strings.TrimSpace(fmt.Sprint(raw["unit"])),
			Quantity:           defaultString(strings.TrimSpace(fmt.Sprint(raw["quantity"])), "0"),
			UnitPrice:          defaultString(strings.TrimSpace(fmt.Sprint(raw["unit_price"])), "0"),
			Amount:             defaultString(strings.TrimSpace(fmt.Sprint(raw["amount"])), "0"),
			SortOrder:          i + 1,
		})
	}
	sort.SliceStable(converted, func(i, j int) bool {
		if converted[i].SortOrder == converted[j].SortOrder {
			return converted[i].Code < converted[j].Code
		}
		return converted[i].SortOrder < converted[j].SortOrder
	})
	optionsJSON, _ := json.Marshal(req.Options)
	sourceJSON, _ := json.Marshal(req.BudgetItems)
	resultJSON, _ := json.Marshal(converted)
	sessionID := fmt.Sprintf("B2B-%d", time.Now().UTC().UnixNano())
	now := time.Now().UTC().Format(time.RFC3339Nano)
	tx, err := r.store.db.BeginTx(ctx, nil)
	if err != nil {
		return BudgetBidConversionSession{}, err
	}
	defer tx.Rollback()
	if req.Mode == "CREATE" {
		var existing string
		err = tx.QueryRowContext(ctx, `SELECT id FROM budget_bid_conversion_sessions WHERE target_bid_project_code=? LIMIT 1`, req.TargetBidProjectCode).Scan(&existing)
		if err == nil {
			return BudgetBidConversionSession{}, errx.New(errx.CodeConflict, "target bid project already has a conversion", "P4-CONV-005")
		}
		if err != sql.ErrNoRows {
			return BudgetBidConversionSession{}, err
		}
	}
	_, err = tx.ExecContext(ctx, `INSERT INTO budget_bid_conversion_sessions(id,source_project_code,source_budget_version_id,target_bid_project_code,mode,status,options_json,source_snapshot_json,result_snapshot_json,created_by,created_at,row_version) VALUES(?,?,?,?,?,'COMPLETED',?,?,?,?,?,1)`,
		sessionID, req.SourceProjectCode, req.SourceBudgetVersionID, req.TargetBidProjectCode, req.Mode, string(optionsJSON), string(sourceJSON), string(resultJSON), req.ActorID, now)
	if err != nil {
		return BudgetBidConversionSession{}, err
	}
	if err = tx.Commit(); err != nil {
		return BudgetBidConversionSession{}, err
	}
	return r.Get(ctx, sessionID)
}

func (r *BudgetBidConversionRepository) Get(ctx context.Context, id string) (BudgetBidConversionSession, error) {
	var item BudgetBidConversionSession
	var optionsJSON, sourceJSON, resultJSON string
	err := r.store.db.QueryRowContext(ctx, `SELECT id,source_project_code,source_budget_version_id,target_bid_project_code,mode,status,options_json,source_snapshot_json,result_snapshot_json,created_by,created_at,row_version FROM budget_bid_conversion_sessions WHERE id=?`, id).
		Scan(&item.ID, &item.SourceProjectCode, &item.SourceBudgetVersionID, &item.TargetBidProjectCode, &item.Mode, &item.Status, &optionsJSON, &sourceJSON, &resultJSON, &item.CreatedBy, &item.CreatedAt, &item.RowVersion)
	if err == sql.ErrNoRows {
		return item, errx.New(errx.CodeNotFound, "conversion session not found", "P4-CONV-005")
	}
	if err != nil {
		return item, err
	}
	_ = json.Unmarshal([]byte(optionsJSON), &item.Options)
	_ = json.Unmarshal([]byte(sourceJSON), &item.SourceSnapshot)
	_ = json.Unmarshal([]byte(resultJSON), &item.ResultSnapshot)
	item.Lineage = map[string]string{"source_project_code": item.SourceProjectCode, "source_budget_version_id": item.SourceBudgetVersionID, "target_bid_project_code": item.TargetBidProjectCode, "session_id": item.ID}
	item.DeepLink = "/app/bid-conversion?session=" + item.ID
	return item, nil
}
